---
title: "Sequence Import and Cleaning"
author:
  - name: Simon Grimm
    url: {simongrimm.com}
date: 2023-02-16
description: |
categories:
draft: false
output:
  bookdown::html_document2:
    toc: yes
    toc_float: yes
    toc_collapsed: yes
    df_print: paged
    code_folding: show
    theme: lumen
    warning: false
    cache: true
    echo: false
    autodep: true
    cache.comment: false
    dpi: 300
editor_options:
  markdown:
    wrap: 72
---

# Initial document dump
[Pooled Amplicon Experiment](https://docs.google.com/document/d/1YRjikGYGGN6P9ZY36nNZMrqFOs_UfweKkipn1_YooO4/edit)

[Github repo with code](https://github.com/naobservatory/phagemid-design/tree/main/analysis/2022-06-06-nanopore-analysis)

[Pooling spradsheet](https://docs.google.com/spreadsheets/d/12b04wXfEUhwDhB-e1jmSayz_7HByKeuQ8hD0rTrcnv4/edit#gid=0)

[More information on phagemid structure](https://docs.google.com/document/d/1qNWIACM-xcg10vNcXyWT7-P4I7DpGqGQ3ywUtDN5qug/edit)

AWS S3 bucket with data: s3://summer-seq-data/230203laura.tar

My task according to Mike:
Data analysis that will be done by Jeff and/or Simon: Compare the measured concentration of each qPCR barcode in two different ways â€” directly by qPCR of that barcode, and indirectly by summing up the estimated concentrations of the 5 corresponding sequencing barcodes.

TODOS:
 - Concatenate data
 - align data with primers given by mike
  		- Forward primer: AGCTTCGGGCCTAACTAACTACT
		- Reverse primer: CGCTTTACGTTACGTCGAAAGC
		Mike: "this should be the primers used. You can confirm by loading the Nanopore data into R and looking for the matching sequences., and more generally by aligning the nanopore reads and the primer sequences"
 - Reach out to anjali to check in on length
 - Check in with Mike to ask how sequencing worked


# Introduction
Here we import, format, and clean

# Setup

## R packages

```{r, message=FALSE}
# set of data analysis and viz tools
library(tidyverse)

# parallel processing
library(furrr)

# file system helpers
library(fs)
library(googlesheets4)

# specifying locations within a project
library(here)

#Fastq file manipulation
import::from(Biostrings, complement, reverseComplement, width, DNAString,
  DNAStringSet, readDNAStringSet, subseq,
  vmatchPDict, vcountPDict
)

import::from(DECIPHER, TrimDNA, AlignSeqs, OrientNucleotides, BrowseSeqs)

# plotting helpers
library(cowplot)
theme_set(theme_cowplot())
library(patchwork)
library(ggbeeswarm)
```

## Defining directories

```{r, message=FALSE}
metadata_data_path <- here("data/metadata")
taxonomy_data_path <- here("data/taxonomy")
output_data_path <- here("output")
project_path <- here('../illumina_pilot')
```

```{r}
#Reading in sequencing data

```